File: backend-node/src/api/rendering.ts
/**
 * JSON rendering with exact control over how each scalar reaches the wire.
 *
 * `JSON.stringify` will not do. Money must arrive as a bare JSON number that
 * keeps the scale it carries (1250.00, not 1250 and not "1250.00"), which no
 * JavaScript number can represent -- 1250.00 and 1250 are the same value. So
 * the document is written directly, and a Money renders from its own digits.
 */

import { EnumValue } from '../domain/enums.js'
import { Instant } from '../domain/instant.js'
import { Money } from '../domain/money.js'

function write(value: unknown, out: string[]): void {
  if (value === null || value === undefined) {
    out.push('null')
    return
  }
  if (value instanceof Money) {
    out.push(value.toString())
    return
  }
  if (value instanceof EnumValue) {
    // An ordinal naming no member is written back as a bare number, exactly as
    // JsonStringEnumConverter does.
    out.push(value.isDefined ? JSON.stringify(value.wireName) : String(value.ordinal))
    return
  }
  if (value instanceof Instant) {
    out.push(JSON.stringify(value.toString()))
    return
  }
  if (value instanceof Date) {
    out.push(JSON.stringify(Instant.fromDate(value).toString()))
    return
  }

  switch (typeof value) {
    case 'boolean':
      out.push(value ? 'true' : 'false')
      return
    case 'number':
      out.push(Number.isFinite(value) ? String(value) : 'null')
      return
    case 'bigint':
      out.push(value.toString())
      return
    case 'string':
      out.push(JSON.stringify(value))
      return
    default:
      break
  }

  if (Array.isArray(value)) {
    out.push('[')
    value.forEach((item, index) => {
      if (index > 0) out.push(',')
      write(item, out)
    })
    out.push(']')
    return
  }

  out.push('{')
  let first = true
  for (const [key, item] of Object.entries(value as Record<string, unknown>)) {
    if (item === undefined) continue
    if (!first) out.push(',')
    first = false
    out.push(JSON.stringify(key), ':')
    write(item, out)
  }
  out.push('}')
}

export function renderJson(payload: unknown): string {
  const out: string[] = []
  write(payload, out)
  return out.join('')
}

File: backend-rb/lib/api/rendering.rb
# frozen_string_literal: true

module Api
  # JSON rendering with exact control over how each scalar reaches the wire.
  #
  # The stock generator will not do. Money must arrive as a *bare* JSON number
  # that keeps the scale it carries (1250.00, not 1250.0 and not "1250.00"),
  # which no Ruby Float can represent. So the document is written directly, and
  # a Money renders from its own digits.
  module Rendering
    module_function

    def dump(payload)
      out = +""
      write(payload, out)
      out
    end

    def write(value, out)
      case value
      when nil then out << "null"
      when true then out << "true"
      when false then out << "false"
      when Domain::Money then out << value.to_s
      when Domain::Enums::Value
        # An ordinal naming no member is written back as a bare number, exactly
        # as JsonStringEnumConverter does.
        out << (value.defined? ? value.wire_name.to_json : value.ordinal.to_s)
      when Domain::Instant then out << value.to_s.to_json
      when ::Integer then out << value.to_s
      when ::String then out << value.to_json
      when ::Time then out << Domain::Instant.from_time(value).to_s.to_json
      when ::Array then write_array(value, out)
      when ::Hash then write_hash(value, out)
      else out << value.to_json
      end
    end

    def write_array(value, out)
      out << "["
      value.each_with_index do |item, index|
        out << "," if index.positive?
        write(item, out)
      end
      out << "]"
    end

    def write_hash(value, out)
      out << "{"
      first = true
      value.each do |key, item|
        out << "," unless first
        first = false
        out << key.to_s.to_json << ":"
        write(item, out)
      end
      out << "}"
    end
  end
end
